### PR TITLE
Basic inline asm support

### PR DESCRIPTION
**Pull request by bjorn3**
You need to use fixed registers. Register classes are not yet supported.

### CODE REVIEW

File: src/inline_asm.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    let mut generated_asm = String::new();
+    writeln!(generated_asm, ".globl {}", asm_name).unwrap();
+    writeln!(generated_asm, "{}:", asm_name).unwrap();
+
+    generated_asm.push_str(".intel_syntax noprefix\n");
+    generated_asm.push_str("    push rbp\n");
+    generated_asm.push_str("    mov rbp,rdi\n");
+    if !options.contains(InlineAsmOptions::PRESERVES_FLAGS) {
+        generated_asm.push_str("    pushf\n");
+    }
```
Comment (Amanieu) [inline src/inline_asm.rs:89]: Flags are not preserved by the calling convention so you don't need to save them here.

File: src/inline_asm.rs [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        return;
+    }
+
+    let mut slot_size = Size::from_bytes(0);
+    let mut clobbered_regs = Vec::new();
+    let mut inputs = Vec::new();
+    let mut outputs = Vec::new();
+
+    let mut new_slot = |reg_class| {
+        // FIXME align slot
```
Comment (Amanieu) [inline src/inline_asm.rs:29]: You can assume that all register sizes are powers of 2 and that they are aligned to their size.

File: src/inline_asm.rs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    writeln!(generated_asm, "{}:", asm_name).unwrap();
+
+    generated_asm.push_str(".intel_syntax noprefix\n");
+    generated_asm.push_str("    push rbp\n");
+    generated_asm.push_str("    mov rbp,rdi\n");
+    if !options.contains(InlineAsmOptions::PRESERVES_FLAGS) {
+        generated_asm.push_str("    pushf\n");
+    }
+
+    // Save clobbered registers
```
Comment (Amanieu) [inline src/inline_asm.rs:91]: You should only save clobbered registers that are required to be saved by the calling convention.

File: src/inline_asm.rs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+fn expect_reg(reg_or_class: InlineAsmRegOrRegClass) -> InlineAsmReg {
+    match reg_or_class {
+        InlineAsmRegOrRegClass::Reg(reg) => reg,
+        InlineAsmRegOrRegClass::RegClass(class) => unimplemented!("{:?}", class),
+    }
+}
+
+fn reg_class_size(reg_class: InlineAsmRegClass) -> Size {
```
Comment (Amanieu) [inline src/inline_asm.rs:193]: I think you can actually derive this from `reg_class.supported_types()`. Just take the largest size of all supported types for the register class.

File: src/inline_asm.rs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    asm_name: &str,
+    arch: InlineAsmArch,
+    options: InlineAsmOptions,
+    template: &[InlineAsmTemplatePiece],
+    clobbered_regs: Vec<(InlineAsmReg, Size)>,
+    inputs: &[(InlineAsmReg, Size, Value)],
+    outputs: &[(InlineAsmReg, Size, CPlace<'_>)],
+) -> String {
+    let mut generated_asm = String::new();
+    writeln!(generated_asm, ".globl {}", asm_name).unwrap();
```
Comment (Amanieu) [inline src/inline_asm.rs:81]: You want some additional metadata here, have a look at what rustc generates. In particular:
- You should create a section just for this function so the linker can discard it if it is not used. (`.section .text.foo,"ax",@progbits`). Use `.text` to switch back afterwards.
- Mark the symbol as a function type (`.type foo,@function`)
- Set the size of the symbol (`.size foo, foo_end-foo`)

File: src/inline_asm.rs [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+                let reg = expect_reg(reg);
+                clobbered_regs.push((reg, new_slot(reg.reg_class())));
+                inputs.push((reg, new_slot(reg.reg_class()), crate::base::trans_operand(fx, in_value).load_scalar(fx)));
+                if let Some(out_place) = out_place {
+                    outputs.push((reg, new_slot(reg.reg_class()), crate::base::trans_place(fx, out_place)));
+                }
+            }
+            InlineAsmOperand::Const { value: _ } => todo!(),
+            InlineAsmOperand::SymFn { value: _ } => todo!(),
+            InlineAsmOperand::SymStatic { def_id: _ } => todo!(),
```
Comment (Amanieu) [inline src/inline_asm.rs:59]: Is there a reason why you're not going through librustc_codegen_ssa's `FunctionCx::codegen_terminator`? `codegen_asm_terminator` does a bunch of necessary lowering for `Const`, `SymFn` and `SymStatic` that you will need to support them.

Comment (bjorn3) [reply src/inline_asm.rs:59]: I don't yet use `rustc_codegen_ssa` for anything but the linking code. `fx` is of type `crate::common::FunctionCx`, not `rustc_codegen_ssa::mir::FunctionCx`.

File: src/inline_asm.rs (hunk 2 above)
Comment (bjorn3) [reply src/inline_asm.rs:29]: `reg_byte` registers have a size of 1. If you first have a `reg_byte` register and then a `reg` register then the `reg` register is misaligned.

Comment (Amanieu) [reply src/inline_asm.rs:29]: If you assume that align == size then you can just round up the offset to a multiple of the size.

Comment (bjorn3) [reply src/inline_asm.rs]: I misunderstood your comment. I thought you meant that the offset was already aligned in all cases. I have now implemented the align logic.

File: src/inline_asm.rs (hunk 3 above)
Comment (bjorn3) [reply src/inline_asm.rs:97]: Added a fixme.

File: src/inline_asm.rs [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+        generated_asm.push_str("    pop rbp\n");
+        generated_asm.push_str("    ret\n");
+    } else {
+        generated_asm.push_str("    ud2\n");
+    }
+
+    generated_asm.push_str(".att_syntax\n");
+    writeln!(generated_asm, ".size {name}, .-{name}", name=asm_name).unwrap();
+    generated_asm.push_str(".popsection\n");
```
Comment (Amanieu) [inline src/inline_asm.rs:148]: It's safer to just reset the section with `.text` here.

File: src/inline_asm.rs [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    _span: Span,
+    template: &[InlineAsmTemplatePiece],
+    operands: &[InlineAsmOperand<'tcx>],
+    options: InlineAsmOptions,
+) {
+    // FIXME add .eh_frame unwind info directives
+
+    if template.is_empty() {
+        // Black box
+        return;
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply src/inline_asm.rs:20]: I know. This is just to not error when using `black_box` on macOS and Windows. I haven't yet implemented assembly support for them. It doesn't matter that much right now as Cranelift barely optimizes things right now.

File: src/inline_asm.rs [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+            .iter()
+            .map(|(ty, _)| ty.size())
+            .max()
+            .unwrap();
+        let align = rustc_target::abi::Align::from_bytes(reg_size.bytes()).unwrap();
+        slot_size = slot_size.align_to(align);
+        let offset = slot_size;
+        slot_size += reg_size;
+        offset
+    };
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply src/inline_asm.rs:40]: Added fixme

**Pull request merged**